### PR TITLE
Pass control values to the L-Qubit kernels

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
Following up on #516 , this PR introduces `controlled_values` everywhere to improve the flexibility of the n-control interface. 

**Description of the Change:**
Pass `controlled_values` around everywhere `controlled_wires` are to be found. I also remove the `extern` definitions from `GateImplementationsLM.hpp` (not effect on build times AFAIK).

**Benefits:**
Simpler decompositions for non-strictly `1`-controlled gates (improved perf). For example, this is important in Grover's algorithm. While it makes no difference when searching for the `111...111` string (best case scenario), it makes a significant difference (~50%) when searching for the `000...000` string (worst case scenario). These are extreme cases, but it gives an idea. Here are some benchmarks performed on a single Perlmutter core before and after this PR.
time_vs_nqubits

The `extern` definitions need not be maintained.

**Possible Drawbacks:**
Increased code complexity.

**Related GitHub Issues:**

### CODE REVIEW

**Comment by vincentmr** [comment]
Support arbitrary controlled values in lightning [sc-51996]

File: pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaManaged.hpp [hunk 1]
```diff
@@ -297,11 +298,15 @@ class StateVectorCudaManaged
...
     applyOperation(const std::string &opName,
                    const std::vector<std::size_t> &controlled_wires,
+                   const std::vector<bool> &controlled_values,
                    const std::vector<std::size_t> &wires, bool inverse = false,
                    const std::vector<Precision> &params = {0.0},
                    const std::vector<complex_t<Precision>> &gate_matrix = {}) {
-        PL_ABORT_IF(!controlled_wires.empty(),
-                    "Controlled kernels not implemented.");
+        PL_ABORT_IF_NOT(controlled_wires.empty(),
+                        "Controlled kernels not implemented.");
```
Comment (AmintorDusko) [inline pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaManaged.hpp:306]: Nice!

File: pennylane_lightning/lightning_qubit/lightning_qubit.py [hunk 2]
```diff
@@ -374,25 +374,26 @@ def _apply_lightning_controlled(self, sim, operation):
...
                 target_wires = self.wires.indices(operation.target_wires)
-            paulix = getattr(sim, "PauliX", None)
-            for wire, value in zip(control_wires, control_values):
-                if not value:
-                    paulix([wire], False, [])
             if method is not None:  # apply n-controlled specialized gate
                 inv = False
                 param = operation.parameters
-                method(control_wires, target_wires, inv, param)
+                method(control_wires, control_values, target_wires, inv, param)
```
Comment (AmintorDusko) [inline pennylane_lightning/lightning_qubit/lightning_qubit.py:380]: This line is not been covered by tests. Could you please look into that?

File: pennylane_lightning/lightning_qubit/lightning_qubit.py [hunk 3]
```diff
@@ -374,25 +374,26 @@ def _apply_lightning_controlled(self, sim, operation):
...
             else:  # apply gate as an n-controlled matrix
                 method = getattr(sim, "applyControlledMatrix")
                 target_wires = self.wires.indices(operation.target_wires)
                 try:
-                    method(qml.matrix(operation.base), control_wires, target_wires, False)
+                    method(
```
Comment (AmintorDusko) [inline pennylane_lightning/lightning_qubit/lightning_qubit.py:385]: Codecov is reporting this line too.

**Review by AmintorDusko — CHANGES_REQUESTED** [review]
Nice work. Super close to approval! I left some comments.

File: pennylane_lightning/lightning_qubit/lightning_qubit.py (hunk 2 above)
Comment (vincentmr) [reply pennylane_lightning/lightning_qubit/lightning_qubit.py:380]: Yes. [Here](https://app.codecov.io/gh/PennyLaneAI/pennylane-lightning/pull/576/blob/pennylane_lightning/lightning_qubit/lightning_qubit.py#L380) it reports proper coverage. The line should be hit by this [test](https://github.com/PennyLaneAI/pennylane-lightning/blob/add_ctrl_values/tests/test_gates.py#L368).

File: pennylane_lightning/lightning_qubit/lightning_qubit.py (hunk 3 above)
Comment (vincentmr) [reply pennylane_lightning/lightning_qubit/lightning_qubit.py:385]: The line should be hit by this [test](https://github.com/PennyLaneAI/pennylane-lightning/blob/add_ctrl_values/tests/test_gates.py#L299).

File: pennylane_lightning/lightning_qubit/lightning_qubit.py (hunk 2 above)
Comment (AmintorDusko) [reply pennylane_lightning/lightning_qubit/lightning_qubit.py:380]: Thanks!

File: pennylane_lightning/lightning_qubit/lightning_qubit.py (hunk 3 above)
Comment (AmintorDusko) [reply pennylane_lightning/lightning_qubit/lightning_qubit.py:385]: Makes sense!

**Review by AmintorDusko — APPROVED** [review]
Thank you for your nice work!

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp [hunk 4]
```diff
@@ -483,17 +497,20 @@ class StateVectorLQubit : public StateVectorBase<PrecisionT, Derived> {
...
     inline void applyControlledMatrix(
         const ComplexT *matrix, const std::vector<size_t> &controlled_wires,
+        const std::vector<bool> &controlled_values,
         const std::vector<size_t> &wires, bool inverse = false) {
         const auto &dispatcher = DynamicDispatcher<PrecisionT>::getInstance();
         auto *arr = this->getData();
-
         PL_ABORT_IF(wires.empty(), "Number of wires must be larger than 0");
-
+        PL_ABORT_IF_NOT(controlled_wires.size() == controlled_values.size(),
```
Comment (multiphaseCFD) [inline pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:511]: Do we want to also make sure `wires.size()>=controlled_wires.size()`?

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp [hunk 5]
```diff
@@ -483,17 +497,20 @@ class StateVectorLQubit : public StateVectorBase<PrecisionT, Derived> {
      * @param kernel Kernel to run the operation
      * @param matrix Pointer to the array data (in row-major format).
      * @param controlled_wires Control wires.
+     * @param controlled_values Control values (0 or 1).
```
Comment (multiphaseCFD) [inline pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:500]: `true` or `false` here?

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp (hunk 4 above)
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:511]: No, these are independent. We could check that they are disjoint sets, but I think this is better done at the Python layer.

**Comment by multiphaseCFD** [comment]
Nice job @vincentmr ! Happy to approve. A kind reminder: please make sure all CI checks passed before merge.

**Review by multiphaseCFD — APPROVED** [review]

**Pull request merged**